C++/Navio/Common/gpio.cpp

```diff
@@ -137,37 +131,37 @@ void Pin::toggle()
     write(!read());
 }
 
+int Pin::getPeriBase() const
+{
+    int rpi_version = getRaspberryPiVersion();
+
+    switch (rpi_version) {
+    case 2:
+    case 3:
+        return BCM2708_PERI_BASE;
+    case 4:
+        return BCM2711_PERI_BASE;
+    default:
+        warnx("Unexpected version. Set PERI_BASE to default as for RPI2/3\n");
+        return BCM2708_PERI_BASE;
+    }
+}
+
 int Pin::getRaspberryPiVersion() const
 {
     char buffer[MAX_SIZE_LINE];
-    const char* hardware_description_entry = "Hardware";
-    const char* v1 = "BCM2708";
-    const char* v2 = "BCM2709";
-    const char* v3 = "BCM2835";
-    char* flag;
-    FILE* fd;
-
-    fd = fopen("/proc/cpuinfo", "r");
-
-    while (fgets(buffer, MAX_SIZE_LINE, fd) != NULL) {
-        flag = strstr(buffer, hardware_description_entry);
-
-        if (flag != NULL) {
-            if (strstr(buffer, v3) != NULL) {
-                fclose(fd);
-                return 3;
-            } else if (strstr(buffer, v2) != NULL) {
-                fclose(fd);
-                return 2;
-            } else if (strstr(buffer, v1) != NULL) {
-                fclose(fd);
-                return 1;
-            }
+    int rpi_version;
+
+    FILE *f = fopen(MODELPATH, "r");
+    if (f != 0 && fgets(buffer, MAX_SIZE_LINE, f) != 0) {
+        int ret = sscanf(buffer + RPI_NAME_LENGTH, "%d", &rpi_version);
+        if (ret == EOF) {
+            rpi_version = 0;
+            warnx("Can't read RPI version\n");
         }
     }
 
-    /* defaults to 1 */
-    fprintf(stderr, "Could not detect RPi version, defaulting to 1\n");
-    fclose(fd);
-    return 1;
+    fclose(f);
+
+    return rpi_version;
```

Comment:
We use C++, not C. Thing such as `RPI_NAME_LENGTH` are _**very**_ bad.
I've not tested it a lot, but you can get the idea from the code below. Just create a file 'test.cpp', paste and compile.
```#include <iostream> // 1 2 3
#include <fstream>
#include <string>

int main(){
    std::size_t i = 0, j = 0;
    std::string line;
    std::ifstream f("test.cpp");
    if (f.is_open() && std::getline(f,line)){
	    for ( ; i < line.length(); i++ ){if (isdigit(line[i])) break;}
	    char ch = line[i];
	    j = atoi(&ch);
	    std::cout << j << std::endl;
	    f.close();

	    // compare the version here
	    return 0;
    }
    std::cout << "unable get contents" << std::endl;
    return 1;
}